File: T4/AvdeevaAN/task1.py
#!/usr/bin/python

import random
import time

def bucketSort(arr, n):
    b = [list() for _ in range(n)]
    for i in range(n):
        bi = int(n * arr[i])
        b[bi].append(arr[i])
    for i in range(n):
        b[i].sort()
    idx = 0
    for i in range(n):
        bucket = b[i]
        for j in bucket:
            arr[idx]=j
            idx += 1

def test(size):
    arr = []
    for i in range(size):
        arr.append(random.uniform(0,1.0))

    start_time = time.time()
    print "Sorted array sizeof", size
    bucketSort(arr, size);
    print("--- %s seconds ---" % (time.time() - start_time))
    print(arr)

test(10)

File: T4/AvdeevaAN/task1.cpp
#include<iostream>
#include<algorithm>
#include<vector>
#include<stdio.h>

using namespace std;

void bucketSort(float arr[], int n)
{
    vector<float> b[n];

    for(int i=0; i<n; i++)
    {
        int bi = n*arr[i];
        b[bi].push_back(arr[i]);
    }

    for(int i=0; i<n; i++)
    {
        sort(b[i].begin(), b[i].end());
    }

    int index = 0;
    for(int i=0; i<n; i++)
    {
        for(int j=0; j<b[i].size(); j++)
        {
            arr[index++]=b[i][j];
        }
    }
}

void test(int size)
{
    float *arr = new float[size];
    for(size_t i = 0; i < size; i++)
    {
        arr[i] = (float)rand() / (float)RAND_MAX;
    }

    time_t t1 = time(0);
    bucketSort(arr, size);
    time_t t2 = time(0);
    printf("Array size %d sorted in %d sec\n", size, t2-t1);
    delete[] arr;
}

int main()
{
    test(10);
}
